op-batcher,op-node,batch_decoder: add logging of compression algo

sebastianst opened the pull request:
<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

**Description**

* Adds logging of the used compression algo to the op-batcher and op-node.
    * To get the compression algo info to the right places in op-node, I added a `Batch` wrapper type `batchWithMetadata` that appends the algo to the log context.
* Adds the algo to the batch decoder tool's output data.
* Fixing some naming around compression algos.
* Adds `brotli` as a valid default selector for `brotli-10` compression

**Tests**

Added the default brotli compression algo string to tests. Remaining changes are logging (or the untested `batch_decoder`)

**Additional context**

There's currently no way to know from batcher or node logs what compression algo is actually in use for (de)compression.

sebastianst committed:
op-batcher: add logging of compression algo

sebastianst committed:
add node logging of algo, add brotli default algo

sebastianst commented on op-batcher/batcher/service.go:
```diff
@@ -242,9 +242,10 @@ func (bs *BatcherService) initChannelConfig(cfg *CLIConfig) error {
 		"max_frame_size", cc.MaxFrameSize,
 		"target_num_frames", cc.TargetNumFrames,
 		"compressor", cc.CompressorConfig.Kind,
+		"compresion_algo", cc.CompressorConfig.CompressionAlgo,
+		"batch_type", cc.BatchType,
```
thanks for finding!

sebastianst committed:
fix typos

axelKingsley commented on op-e2e/e2eutils/geth/wait.go:
```diff
@@ -65,7 +66,8 @@ func WaitForTransaction(hash common.Hash, client *ethclient.Client, timeout time
 		if receipt != nil && err == nil {
 			return receipt, nil
-		} else if err != nil && !errors.Is(err, ethereum.NotFound) {
+		} else if err != nil &&
+			!(errors.Is(err, ethereum.NotFound) || strings.Contains(err.Error(), errStrTxIdxingInProgress)) {
```
nit: can you explain why we now also wait for indexing? Seems like the right call but did you hit test issues?

trianglesphere replied:
geth's made tx indexing asynchronous so at startup it's possible to not be able to do the lookup by hash

sebastianst replied:
This made a bunch of tests fail and is due to a recent change in geth. It is affecting teams thorough the ecosystem, see e.g. this discussion: https://github.com/ethereum/go-ethereum/issues/28877

axelKingsley commented on op-node/cmd/batch_decoder/reassemble/reassemble.go:
```diff
@@ -107,8 +107,12 @@ func processFrames(cfg Config, rollupCfg *rollup.Config, id derive.ChannelID, fr
+	var (
+		batches    []derive.Batch
+		batchTypes []int
+		comprAlgos []derive.CompressionAlgo
```
why are we building a list of algos, wouldn't there only ever be one at this point, when processing frames? I guess theoretically there could be a change in algo, but I hadn't expected the decoder to survive that.

trianglesphere replied:
I tend to think so as well, particularly if we are only looking for frames with the same channel id

sebastianst replied:
I just followed the pattern for `batchTypes` as well. There usually should ever only be one. But this is a debugging tool, so if there ever are different ones, we can detect it here.

trianglesphere commented on op-node/cmd/batch_decoder/reassemble/reassemble.go:
```diff
@@ -18,13 +18,14 @@ import (
+	Frames         []FrameWithMetadata      `json:"frames"`
+	Batches        []derive.Batch           `json:"batches"`
+	BatchTypes     []int                    `json:"batch_types"`
+	ComprAlgos     []derive.CompressionAlgo `json:"compr_alogs"`
```
In here there should only be one compression algo per channel

sebastianst replied:
I just followed the pattern for `batchTypes` and thought since this is a debugging tool, we record it for every batch individually. Otherwise, wouldn't we also need to make `batchTypes` a single values?

trianglesphere reviewed (commented):
derivation + logging changes LGTM. Just some Q's about the re-assemble tool

sebastianst committed:
only log compression algo if present

sebastianst commented:
I removed the op-e2e receipt confirmation change, will handle in a separate PR.

sebastianst committed:
add type conversion abstraction to Batch interface

Since we're dealing now with wrapped types around Batch implementations,
this let's us transparently unwrap the underlying batch types. It makes
sense to add this to the interface, because getting the underlying types
from the interface is done in several places, so it's part of the
interface's contract.

sebastianst committed:
adapt BatchReader test
